services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/util/DeviceRegistryUtils.java

```diff
@@ -134,4 +149,48 @@ public static CacheDirective getCacheDirective(final int cacheMaxAge) {
     public static String getUniqueIdentifier() {
         return UUID.randomUUID().toString();
     }
+
+    /**
+     * Gets the certificate of the device to be provisioned from the client context.
+     *
+     * @param tenantId The tenant to which the device belongs.
+     * @param authId The authentication identifier.
+     * @param clientContext The client context that can be used to get the X.509 certificate 
+     *                      of the device to be provisioned.
+     * @param span The active OpenTracing span for this operation. It is not to be closed in this method! An
+     *             implementation should log (error) events on this span and it may set tags and use this span 
+     *             as the parent for any spans created in this method.
+     * @return the certificate of the device to be provisioned.
+     * @throws NullPointerException if any of the parameters except span is {@code null}.
+     * @throws ClientErrorException if error getting the certificate from the given client context.
+     */
+    public static X509Certificate getCertificateFromClientContext(
+            final String tenantId,
+            final String authId,
+            final JsonObject clientContext,
+            final Span span) throws ClientErrorException {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(authId);
+        Objects.requireNonNull(clientContext);
+
+        try {
+            final byte[] bytes = clientContext.getBinary(CredentialsConstants.FIELD_CLIENT_CERT);
```

Comment:
what if the client context doesn't contain a certificate?

Reply:
@sophokles73 I just pulled up the already existing code to a utility class and didn't go through the details inside the method assuming that it has been already checked. How about adding a null check and throwing `IllegalArgumentException` saying that the certificate is not available?

Reply:
that would only make sense if the client has a chance of preventing it to be thrown. For that, it would need to check for the presence of the JSON property, right? That, however, is what we are trying to encapsulate here, isn't it?

How about returning a `Future<Optional<X509Certificate>>` and let the client only process the result if it is not empty?